api4/data_retention.go

```diff
@@ -4,21 +4,294 @@
+	sCountsOnly := r.URL.Query().Get("counts_only")
+	bCountsOnly := false
+	var err error
+	var body map[string]interface{}
+	if sCountsOnly != "" {
+		bCountsOnly, err = strconv.ParseBool(sCountsOnly)
+		if err != nil {
+			mlog.Warn("Failed to parse counts_only URL query parameter from getPolicies request", mlog.Err(err))
+		}
+	}
+	if bCountsOnly {
+		policies, err := c.App.GetRetentionPoliciesWithCounts()
+		if err != nil {
+			c.Err = err
+			return
+		}
+		body = map[string]interface{}{
+			"policies":    policies,
+			"total_count": len(policies),
+		}
+	} else {
+		policies, err := c.App.GetRetentionPolicies()
+		if err != nil {
+			c.Err = err
+			return
+		}
+		body = map[string]interface{}{
+			"policies":    policies,
+			"total_count": len(policies),
+		}
+	}
+	b, _ := json.Marshal(body)
+	w.Write(b)
+}
+
+func getPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policyId := mux.Vars(r)["policy_id"]
+	policy, err := c.App.GetRetentionPolicy(policyId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	w.Write(policy.ToJsonBytes())
+}
+
+func createPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policy, jsonErr := model.RetentionPolicyFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+		return
+	}
+	auditRec := c.MakeAuditRecord("createPolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("policy", policy)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	policy, err := c.App.CreateRetentionPolicy(policy)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	auditRec.AddMeta("policy", policy) // overwrite meta
+	auditRec.Success()
+	w.WriteHeader(http.StatusCreated)
+	w.Write(policy.ToJsonBytes())
+}
+
+func patchPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	patch, jsonErr := model.RetentionPolicyUpdateFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+	}
+	patch.Id = mux.Vars(r)["policy_id"]
+
+	auditRec := c.MakeAuditRecord("patchPolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("patch", patch)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	policy, err := c.App.PatchRetentionPolicy(patch)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	auditRec.Success()
+	w.Write(policy.ToJsonBytes())
+}
+
+func updatePolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	update, jsonErr := model.RetentionPolicyUpdateFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+	}
+	update.Id = mux.Vars(r)["policy_id"]
+
+	auditRec := c.MakeAuditRecord("updatePolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("update", update)
+	if !c.IsSystemAdmin() {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	policy, err := c.App.UpdateRetentionPolicy(update)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	auditRec.Success()
+	w.Write(policy.ToJsonBytes())
+}
```

Comment:
The only difference is that `Update` will check that all necessary struct fields are present in the JSON body. This is done to prevent the client from accidentally resetting some fields to their "default" values, which is what we currently do with some of our other API endpoints.